## Deliverable 2: Determine Bias of Vine Reviews
For the second deliverable, use PySpark, Pandas, or SQL to determine if there is any bias towards reviews that were written as part of the Vine program. For this analysis determine if there are differences in the percentage of 5 star reviews where there is a written Vine review (i.e., paid) compared to where there isn’t a written Vine review (i.e, unpaid).

In [2]:
# Import Dependencies
import pandas as pd

In [3]:
# use pgAdmin to export the vine_table as a csv file 
# this file is has been saved as vine_table.csv in this repository

# Read in the vine_table.csv file
vine_df = pd.read_csv("./Data/vine_table.csv")
vine_df

review_id  star_rating  helpful_votes  total_votes vine  \
0         REAKC26P07MDN          5.0            0.0          0.0    N   
1        R3NU7OMZ4HQIEG          2.0            0.0          1.0    N   
2        R14QJW3XF8QO1P          5.0            0.0          0.0    N   
3        R2HB7AX0394ZGY          5.0            0.0          0.0    N   
4         RGKMPDQGSAHR3          5.0            0.0          0.0    N   
...                 ...          ...            ...          ...  ...   
2643614   R40AVFLMGSX5Q          1.0           32.0         34.0    N   
2643615  R13RP6H1OP0BFS          5.0            3.0          3.0    N   
2643616  R34ITLN8LBTFZN          3.0           48.0         56.0    N   
2643617   R30D2N6TX0OFS          3.0            7.0          7.0    N   
2643618  R2MAGOAOXPFMF1          2.0           17.0         20.0    N   

        verified_purchase  
0                       Y  
1                       Y  
2                       Y  
3                       Y  
4                       Y  
...                   ...  
2643614                 N  
2643615                 N  
2643616                 N  
2643617                 Y  
2643618                 Y  

[2643619 rows x 6 columns]

In [4]:
# explore the dataframe using the info method
vine_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2643619 entries, 0 to 2643618
Data columns (total 6 columns):
 #   Column             Dtype  
---  ------             -----  
 0   review_id          object 
 1   star_rating        float64
 2   helpful_votes      float64
 3   total_votes        float64
 4   vine               object 
 5   verified_purchase  object 
dtypes: float64(3), object(3)
memory usage: 121.0+ MB


## Step 1.
Filter the data and create a new DataFrame or table to retrieve all the rows where the total_votes count is equal to or greater than 20 to pick reviews that are more likely to be helpful and to avoid having division by zero errors later on.


In [7]:
# filter vine_df for all rows where total_votes >= 20

vine_df_20 = vine_df.loc[vine_df["total_votes"] >= 20]
vine_df_20

# note this reduces the total rows from 2,643,619 to 39,376

review_id  star_rating  helpful_votes  total_votes vine  \
128      R21KC552Y6HL8X          1.0           27.0         31.0    N   
161       RX9WC9FTIR1XR          5.0           25.0         25.0    N   
256       RGDCOU1KBHMNG          3.0           29.0         31.0    N   
267       RVTYWID2TPMMY          2.0           35.0         42.0    N   
719      R2CMPZ5VESGRLY          4.0           27.0         28.0    N   
...                 ...          ...            ...          ...  ...   
2643586   RDDVO19Y8YH01          1.0           12.0         56.0    N   
2643591   RU2X9GTDZ00IM          5.0           58.0         59.0    N   
2643614   R40AVFLMGSX5Q          1.0           32.0         34.0    N   
2643616  R34ITLN8LBTFZN          3.0           48.0         56.0    N   
2643618  R2MAGOAOXPFMF1          2.0           17.0         20.0    N   

        verified_purchase  
128                     Y  
161                     Y  
256                     Y  
267                     Y  
719                     Y  
...                   ...  
2643586                 N  
2643591                 N  
2643614                 N  
2643616                 N  
2643618                 Y  

[39376 rows x 6 columns]

## step 2.
Filter the new DataFrame or table created in Step 1 and create a new DataFrame or table to retrieve all the rows where the number of helpful_votes divided by total_votes is equal to or greater than 50%.

In [8]:
# filter vine_df_20 where helpful_votes/total_votes >= 0.50
vine_df_50 = vine_df_20.loc[(vine_df_20["helpful_votes"] / vine_df_20["total_votes"]) >= 0.50]
vine_df_50

# this reduces the number of rows from 39,376 to 38,010 

review_id  star_rating  helpful_votes  total_votes vine  \
128      R21KC552Y6HL8X          1.0           27.0         31.0    N   
161       RX9WC9FTIR1XR          5.0           25.0         25.0    N   
256       RGDCOU1KBHMNG          3.0           29.0         31.0    N   
267       RVTYWID2TPMMY          2.0           35.0         42.0    N   
719      R2CMPZ5VESGRLY          4.0           27.0         28.0    N   
...                 ...          ...            ...          ...  ...   
2643577  R37NBT4U1EL04K          5.0           47.0         61.0    N   
2643591   RU2X9GTDZ00IM          5.0           58.0         59.0    N   
2643614   R40AVFLMGSX5Q          1.0           32.0         34.0    N   
2643616  R34ITLN8LBTFZN          3.0           48.0         56.0    N   
2643618  R2MAGOAOXPFMF1          2.0           17.0         20.0    N   

        verified_purchase  
128                     Y  
161                     Y  
256                     Y  
267                     Y  
719                     Y  
...                   ...  
2643577                 N  
2643591                 N  
2643614                 N  
2643616                 N  
2643618                 Y  

[38010 rows x 6 columns]

## step 3. 
Filter the DataFrame or table created in Step 2, and create a new DataFrame or table that retrieves all the rows where a review was written as part of the Vine program (paid), vine == 'Y'.

In [10]:
# filter vine_df_50 where vine = "Y"
paid_reviews_df = vine_df_50.loc[vine_df_50["vine"] == 'Y']
paid_reviews_df

# note there are 170 paid reviews (that meet the filter criteria from steps 1 & 2)

review_id  star_rating  helpful_votes  total_votes vine  \
8547     R3A71VR1JZD8WF          2.0           27.0         30.0    Y   
10246    R16OMUJIGI18JZ          5.0           72.0         72.0    Y   
25168    R3TS8ZP2FHQ9XR          5.0           39.0         42.0    Y   
46422    R2MHP919VZN7DI          5.0           29.0         30.0    Y   
66446     RD2BCTVS59A5L          2.0           20.0         20.0    Y   
...                 ...          ...            ...          ...  ...   
2596353   RV5OQXGCYHSC6          5.0           24.0         28.0    Y   
2596519  R2XLAQL0E2UPWQ          4.0           33.0         34.0    Y   
2597336  R2UJFVH97OJORD          3.0          106.0        117.0    Y   
2597830   RJXG84I331N9T          3.0          146.0        154.0    Y   
2598196   RNGRV0353CDWX          4.0          109.0        115.0    Y   

        verified_purchase  
8547                    N  
10246                   N  
25168                   N  
46422                   N  
66446                   N  
...                   ...  
2596353                 N  
2596519                 N  
2597336                 N  
2597830                 N  
2598196                 N  

[170 rows x 6 columns]

## step 4
Repeat Step 3, but this time retrieve all the rows where the review was not part of the Vine program (unpaid), vine == 'N'.

In [11]:
# filter vine_df_50 where vine = "N"
unpaid_reviews_df = vine_df_50.loc[vine_df_50["vine"] == 'N']
unpaid_reviews_df

# note there are 37840 unpaid reviews (that meet the filter criteria from steps 1 & 2)

review_id  star_rating  helpful_votes  total_votes vine  \
128      R21KC552Y6HL8X          1.0           27.0         31.0    N   
161       RX9WC9FTIR1XR          5.0           25.0         25.0    N   
256       RGDCOU1KBHMNG          3.0           29.0         31.0    N   
267       RVTYWID2TPMMY          2.0           35.0         42.0    N   
719      R2CMPZ5VESGRLY          4.0           27.0         28.0    N   
...                 ...          ...            ...          ...  ...   
2643577  R37NBT4U1EL04K          5.0           47.0         61.0    N   
2643591   RU2X9GTDZ00IM          5.0           58.0         59.0    N   
2643614   R40AVFLMGSX5Q          1.0           32.0         34.0    N   
2643616  R34ITLN8LBTFZN          3.0           48.0         56.0    N   
2643618  R2MAGOAOXPFMF1          2.0           17.0         20.0    N   

        verified_purchase  
128                     Y  
161                     Y  
256                     Y  
267                     Y  
719                     Y  
...                   ...  
2643577                 N  
2643591                 N  
2643614                 N  
2643616                 N  
2643618                 Y  

[37840 rows x 6 columns]

## step 5
Determine the total number of reviews, the number of 5-star reviews, and the percentage of 5-star reviews for the two types of review (paid vs unpaid).

### Step 5a. Paid Reviews

In [26]:
# determine the count of paid 5 star reviews 
paid_5star_reviews = paid_reviews_df.loc[paid_reviews_df["star_rating"] == 5]["star_rating"].count()
paid_5star_reviews

65

In [27]:
# determine the count of paid reviews
paid_reviews = paid_reviews_df["star_rating"].count()
paid_reviews

170

In [29]:
# determine the percentage of paid_5star_reviews
# percentage_paid_5star_reviews = paid_5star_reviews / paid_reviews
percentage_paid_5star_reviews = round((paid_5star_reviews / paid_reviews) * 100, 2)
percentage_paid_5star_reviews

38.24

### Step 5b. UnpPaid Reviews

In [30]:
# determine the count of paid 5 star reviews 
unpaid_5star_reviews = unpaid_reviews_df.loc[unpaid_reviews_df["star_rating"] == 5]["star_rating"].count()
unpaid_5star_reviews

20612

In [31]:
# determine the count of unpaid reviews
unpaid_reviews = unpaid_reviews_df["star_rating"].count()
unpaid_reviews

37840

In [32]:
# determine the percentage of unpaid_5star_reviews
# percentage_unpaid_5star_reviews = unpaid_5star_reviews / unpaid_reviews
percentage_unpaid_5star_reviews = round((unpaid_5star_reviews / unpaid_reviews) * 100, 2)
percentage_unpaid_5star_reviews

54.47

## Analysis:
Ho: 5star_vine = 5star_not_vine
(no difference in the percentage (proportion) of 5 star ratings paid (vine) vs unpaid (not vine)

Ha: 5star_vine <> 5star_not_vine
(there is a difference)

* percentage_5star_vine     = 38.24%
* percentage_5star_not_vine = 54.47%

* **finding: The percentage of 5star ratings is not the same when comparing vine reviews to not_vine reviews**
* **There is reason to believe that the percentage of 5star ratings is not directly tied to whether the reviewer is paid or unpaid.**

* note: this hypothesis test would need to be evaluated comparing the proportion of 5 star reviews among the two samples